### ISSUE TITLE
Does not look native on Windows 11

### ISSUE THREAD

**Issue by jhi2** [beeware/toga#2574]
### What is the problem or limitation you are having?

It does not look native on windows 11.
A windows 11 app looks like this:
![image](https://github.com/beeware/toga/assets/116117148/73df1c96-50b3-4edd-8395-8ddb8573e812)

### Describe the solution you'd like

Use winUI3 when on Windows 11

### Describe alternatives you've considered

nativesvttk and tkinter

**Comment by freakboy3742** [comment]
Thanks for the suggestion - we'd love to add a WinUI3 backend; the problem is that [there are no Python bindings for WinUI3](https://github.com/microsoft/microsoft-ui-xaml/issues/3364). Once those bindings exist and are stable, a WinUI3 backend for Toga can be developed.

**Comment by jhi2** [comment]
Maybe I could help make one

**Comment by JPHutchins** [comment]
Hi @jhi2, I am a C programmer and would be happy to help.  In this case, I think that the ctypes + DLL approach taken by win32more makes sense, at least to an MVP.  Generally, I am having a hard time understanding that repo, it is very complex.  It sorta looks like a code generation tool PLUS some generated code, which makes usage confusing.  Plus, organizing all these namespaces by way of directories is perhaps nice for repo organization but it may be not so great for Python import initialization.  I have concerns about the "DLL function decorator".  Perhaps win32more can be used to generate only the bindings that a Toga backend would need?

Happy to help with testing or other tasks.

TBH, anything other than a Microsoft supported solution will be a bit of a hack.  MS probably doesn't have the money or talent for this 🙄.

**Comment by JPHutchins** [comment]
To see how simple the ctypes APIs can be, try this example:

```python
import ctypes as c

user32 = c.WinDLL("user32.dll")

user32.MessageBoxW(0, "Hello World", "Hello", 1)
```

Where the documentation is from here: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-messageboxw

The tedious bit is having a nice wrapper with typing.

I am looking into the winrt package collection: https://pypi.org/search/?q=winrt&page=1

So far I have not been able to reproduce a simpler version of what win32more is doing.  I am able to import an XAML Application class but cannot create the application.

**Comment by JPHutchins** [comment]
The win32more interface to `Application` seems different than WinRT: https://github.com/ynkdir/py-win32more/blob/87f331b014fd1cda3ede67c4cf10715acafe575a/win32more/Microsoft/UI/Xaml/__init__.py#L67-L130

Compare to the interface generated by WinRT:
```python
@typing.final
class Application_Static(type):
    @typing.overload
    def load_component(cls, component: typing.Optional[winrt.system.Object], resource_locator: typing.Optional[windows_foundation.Uri], /) -> None: ...
    @typing.overload
    def load_component(cls, component: typing.Optional[winrt.system.Object], resource_locator: typing.Optional[windows_foundation.Uri], component_resource_location: microsoft_ui_xaml_controls_primitives.ComponentResourceLocation, /) -> None: ...
    def start(cls, callback: typing.Optional[ApplicationInitializationCallback], /) -> None: ...
    @_property
    def current(cls) -> typing.Optional[Application]: ...

@typing.final
class Application(winrt.system.Object, metaclass=Application_Static):
    @staticmethod
    def _from(obj: winrt.system.Object, /) -> Application: ...
    def __new__(cls: typing.Type[Application]) -> Application: ...
    def exit(self) -> None: ...
    def add_unhandled_exception(self, handler: typing.Optional[UnhandledExceptionEventHandler], /) -> windows_foundation.EventRegistrationToken: ...
    def remove_unhandled_exception(self, token: windows_foundation.EventRegistrationToken, /) -> None: ...
    def add_resource_manager_requested(self, handler: windows_foundation.TypedEventHandler[winrt.system.Object, ResourceManagerRequestedEventArgs], /) -> windows_foundation.EventRegistrationToken: ...
    def remove_resource_manager_requested(self, token: windows_foundation.EventRegistrationToken, /) -> None: ...
    @_property
    def resources(self) -> typing.Optional[ResourceDictionary]: ...
    @resources.setter
    def resources(self, value: typing.Optional[ResourceDictionary]) -> None: ...
    @_property
    def requested_theme(self) -> ApplicationTheme: ...
    @requested_theme.setter
    def requested_theme(self, value: ApplicationTheme) -> None: ...
    @_property
    def high_contrast_adjustment(self) -> ApplicationHighContrastAdjustment: ...
    @high_contrast_adjustment.setter
    def high_contrast_adjustment(self, value: ApplicationHighContrastAdjustment) -> None: ...
    @_property
    def focus_visual_kind(self) -> FocusVisualKind: ...
    @focus_visual_kind.setter
    def focus_visual_kind(self, value: FocusVisualKind) -> None: ...
    @_property
    def debug_settings(self) -> typing.Optional[DebugSettings]: ...
    @_property
    def dispatcher_shutdown_mode(self) -> DispatcherShutdownMode: ...
    @dispatcher_shutdown_mode.setter
    def dispatcher_shutdown_mode(self, value: DispatcherShutdownMode) -> None: ...
```

Here's the C# doc: https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.application?view=windows-app-sdk-1.5

**Comment by freakboy3742** [comment]
If you're calling APIs in user32, you're not calling WinUI3 API; you're calling the 1995 era win32 API. It's also available using the win32api package, which is mostly a set of convenience wrappers around the win32 API.

It *may* be possible to use ctypes to call WinUi3 libraries, but the API calls and libraries will be different.

**Comment by JPHutchins** [comment]
> It _may_ be possible to use ctypes to call WinUi3 libraries, but the API calls and libraries will be different.

Yes!  The intention is to show how easy it is to make calls to any DLL from python. 

Presently I'm trying to track down how win32more is hooking into the new APIs.  This decorator is on Application::Start:

https://github.com/ynkdir/py-win32more/blob/87f331b014fd1cda3ede67c4cf10715acafe575a/win32more/_winrt.py#L710-L724

It's getting the "Start" function from this:

https://github.com/ynkdir/py-win32more/blob/87f331b014fd1cda3ede67c4cf10715acafe575a/win32more/_winrt.py#L786-L793

Seems like the magic is in the RoGetActivationFactory, which I've not heard of before: https://learn.microsoft.com/en-us/windows/win32/api/roapi/nf-roapi-rogetactivationfactory

**Comment by JPHutchins** [comment]
```python
xaml = ctypes.WinDLL("windows.ui.xaml.dll")
```

Is actually loading something... not  sure what - can't find any function pointers.

**Comment by freakboy3742** [comment]
If you're digging into this, I'd pay particular attention to [this comment](https://github.com/microsoft/microsoft-ui-xaml/issues/3364#issuecomment-2154514281) on the WinUI3 ticket I've referenced earlier in this discussion. @zooba  is a Microsoft employee and member of the CPython core team; his offer around co-developing bindings should not be ignored or taken lightly.

**Comment by zooba** [comment]
Yeah, this is a _really_ deep hole to dive into :) I was lucky to spend my intern project at MSFT (in 2011) working on this stuff while the APIs were being developed, and I don't think any of the low-level intro material we had ever made it public.

Basically, all the WinRT APIs are approx. COM, which means you don't load them as normal C APIs. The `RoGetActivationFactory`[^1] returns an interface that has function pointers to functions that will create instances of particular objects. Those then have interfaces to access and use those objects (as well as reference counting). It's a very powerful model, particularly for forwards and backwards _binary_ compatibility - something of a holy grail in OS design - but it takes a bit to get it all into your head.

WinUI3 is based around these same APIs, but are registered somewhat differently. I haven't dug too far into all the details, but it's more convenient (and more wasteful of memory and disk space) than the original Windows 8 UWP APIs.

But having now implemented 4 different approaches of accessing it from Python, the one I think is best is to just generate C++ code to do all the GUI side and have it call back into Python. I've implemented something along these lines [here](https://github.com/zooba/pymsbuild-winui)[^2]. It was pretty easy to translate some existing C# samples into Python, but I haven't really tried it at scale (though it already scales so much better than previous efforts that I'm sure it'd be fine). There might be something of value in there for you.

[^1]: IIRC, "RO" stood for Runtime Object. The equivalent COM APIs start with "CO".

[^2]: Based on my [`pymsbuild`](https://github.com/zooba/pymsbuild) build backend, since building is an unavoidable part of code generation.

**Comment by jhi2** [comment]
> Hi @jhi2, I am a C programmer and would be happy to help. In this case, I think that the ctypes + DLL approach taken by win32more makes sense, at least to get an MVP. Generally, I am having a hard time understanding that repo, it is very complex. It sorta looks like a code generation tool PLUS some generated code, which makes usage confusing. Plus, organizing all these namespaces by way of directories is perhaps nice for repo organization but it may be not so great for Python import initialization. I have concerns about the "DLL function decorator". Perhaps win32more can be used to generate only the bindings that a Toga backend would need?
> 
> Happy to help with testing or other tasks.
> 
> TBH, anything other than a Microsoft supported solution will be a bit of a hack. MS probably doesn't have the money or talent for this 🙄.


I like the "MS probably doesn't have the money or talent for this 🙄" That's why they invented Rectify11 haha.
I have a project on this and I may need to make my own binding. Your help would be appreciated! see:https://github.com/jhi2/PyUi3 NOTE:I only know core C++.

**Comment by jhi2** [comment]
`windows.ui.xaml.dll` is for WinForms or WPF `Microsoft.Ui.Xaml `is for WinUI

**Comment by jhi2** [comment]
> Yeah, this is a _really_ deep hole to dive into :) I was lucky to spend my intern project at MSFT (in 2011) working on this stuff while the APIs were being developed, and I don't think any of the low-level intro material we had ever made it public.
> 
> Basically, all the WinRT APIs are approx. COM, which means you don't load them as normal C APIs. The `RoGetActivationFactory`[1](#user-content-fn-1-1769afd4d3f60435aa6e5065c24dfdb1) returns an interface that has function pointers to functions that will create instances of particular objects. Those then have interfaces to access and use those objects (as well as reference counting). It's a very powerful model, particularly for forwards and backwards _binary_ compatibility - something of a holy grail in OS design - but it takes a bit to get it all into your head.
> 
> WinUI3 is based around these same APIs, but are registered somewhat differently. I haven't dug too far into all the details, but it's more convenient (and more wasteful of memory and disk space) than the original Windows 8 UWP APIs.
> 
> But having now implemented 4 different approaches of accessing it from Python, the one I think is best is to just generate C++ code to do all the GUI side and have it call back into Python. I've implemented something along these lines [here](https://github.com/zooba/pymsbuild-winui)[2](#user-content-fn-2-1769afd4d3f60435aa6e5065c24dfdb1). It was pretty easy to translate some existing C# samples into Python, but I haven't really tried it at scale (though it already scales so much better than previous efforts that I'm sure it'd be fine). There might be something of value in there for you.
> 
> ## Footnotes
> 1. IIRC, "RO" stood for Runtime Object. The equivalent COM APIs start with "CO". [↩](#user-content-fnref-1-1769afd4d3f60435aa6e5065c24dfdb1)
> 2. Based on my [`pymsbuild`](https://github.com/zooba/pymsbuild) build backend, since building is an unavoidable part of code generation. [↩](#user-content-fnref-2-1769afd4d3f60435aa6e5065c24dfdb1)

I will use cTypes

**Comment by JPHutchins** [comment]
> I like the "MS probably doesn't have the money or talent for this 🙄" That's why they invented Rectify11 haha.

Literally, cannot wait until AI is "over".  But then there will be the "next thing" that takes resources away from luxuries like being able to produce and distribute signed applications to users.

> `windows.ui.xaml.dll` is for WinForms or WPF `Microsoft.Ui.Xaml `is for WinUI

Will take a look!

Also, consider testing this as the backend instead of win32more: https://github.com/zooba/pymsbuild-winui

**Comment by JPHutchins** [comment]
Very __maybe__ we can get ctypes - I'll have to fiddle.

```python
>>> com = ctypes.WinDLL("combase.dll")
>>> com.RoGetActivationFactory
<_FuncPtr object at 0x000001F9BE423790>
```

**Comment by JPHutchins** [comment]
No, it still seems foolish to try: https://learn.microsoft.com/en-us/windows/win32/api/roapi/nf-roapi-rogetactivationfactory#syntax

> The `RoGetActivationFactory`[1](#user-content-fn-1-1769afd4d3f60435aa6e5065c24dfdb1) returns an interface that has function pointers to functions that will create instances of particular objects. Those then have interfaces to access and use those objects (as well as reference counting).

**Comment by jhi2** [comment]
I would just want to call C++ code from ctypes. I found a way to call Python from C/C#/C++(Python.h)

**Comment by jhi2** [comment]
> Literally, cannot wait until AI is "over". But then there will be the "next thing" that takes resources away from luxuries like being able to produce and distribute signed applications.

Really? how will u sign them? Linux?

**Comment by JPHutchins** [comment]
> I would just want to call C++ code from ctypes. I found a way to call Python from C/C#/C++(Python.h)

This might be nice, but I'll recommend pymsbuild-winui or win32more for generating the bindings.  I think it might be possible with ctypes but feels hacky.  I mean, it's all hacky.  That RoGetActivationFactory will provide a function pointer, but I'm not sure how to bind it to a python function.  Maybe with this: https://docs.python.org/3/library/ctypes.html#callback-functions

**Comment by zooba** [comment]
You'll actually want [function prototypes](https://docs.python.org/3.13/library/ctypes.html#function-prototypes), but you might also be able to find a COM library that implements the core functionality (e.g. the [`IUnknown`](https://learn.microsoft.com/en-us/windows/win32/api/unknwn/nn-unknwn-iunknown) interface, which is the default function table that every interface starts with).

(And to be clear, I'm only recommending this for research purposes. There's just no way to build everything you'll need with `ctypes`, but you can certainly explore a bit and there's a ton of interesting stuff to learn!)

**Comment by JPHutchins** [comment]
Presently leaning toward a proof of concept that is a PyPI package with static libs based on pymsbuild-winui.

One goal is to avoid developers needing to install the VS toolchains.

**Comment by jhi2** [comment]
> Also, consider testing this as the backend instead of win32more: https://github.com/zooba/pymsbuild-winui
Will try that

**Comment by jhi2** [comment]
I DO NOT LIKE the setup of pymsbuild-winui

**Comment by JPHutchins** [comment]
> I DO NOT LIKE the setup of pymsbuild-winui

It's not setup as a package yet, so that would be the work, AFAICT.  I really hope to have some time to look into this soon!

**Comment by jhi2** [comment]
I have a remarkable new breakthrough, based on the foundation that WinUI3 is a COM object. I will come out with the code soon here.

**Comment by jhi2** [comment]
Here is the code I promised:
```shell
pip install pywin32 
``` 
```python
import win32com.client
Microsoft.Ui.Xaml = win32com.client.Dispatch("Microsoft.Ui.Xaml")
#WinUI3 is now accessable
```

**Comment by jhi2** [comment]
Anyone?

**Comment by freakboy3742** [comment]
Anyone what? What response are you expecting?

You've posted a snippet of code that suggests that the WinUI3 API may be accessible - using an approach that didn't occur to a Microsoft employee who worked on the problem while working at Microsoft. 

Even if we assume that the approach you've described *does* work - there is a *huge* gap between "here's how you can access the WinUI3 libraries" and "Here's a WinUI3 backend for Toga.". It *definitely* didn't happen in the 30 minutes between you posting the proposed solution and responding "Anyone?".

**Comment by jhi2** [comment]
Ok

**Comment by jhi2** [comment]
Can you look it over?

**Comment by JPHutchins** [comment]
> Can you look it over?

I can look it over.  Think of this as an exploratory phase where we're looking for cleanest starts of a minimum viable product.

Ideally, we reproduce this in "pure" python: https://github.com/sotanakamura/winui3-without-xaml

I'll be fiddling presently, so feel free to send examples of what you have so far.

**Comment by jhi2** [comment]
Can't the Toga code generate the xaml?

**Comment by JPHutchins** [comment]
> Can't the Toga code generate the xaml?

That would be one approach.

**Comment by freakboy3742** [comment]
Look over what? You've provided 3 lines of code. 

I'm sceptical that the approach you've described will work - because if something that simple *would* work, I feel like @zooba would have suggested it. However, I have no particular expertise with COM, WinUI3, or Windows in general, so I'm not really in a position to provide any sort of critical analysis of the approach.

Adding a WinUI3 backend isn't on my roadmap for this quarter - nor is it likely to be on my roadmap in the foreseeable future. If this is something that interests *you*, than I'm able to provide advice on how to go about building a WinUI3 backend; and I'm willing to review code in the form of PRs against the Toga repo - but until there's working WinUI3 code that can run, there's not much more I'm in a position to offer.

And - answering your follow up question - no, Toga can't use a XAML approach. Toga needs to be able to programmatically drive the construction of widgets. Generating markup isn't an approach that is fundamentally compatible with Toga.

**Comment by jhi2** [comment]
> > Can you look it over?
> 
> I can look it over. Think of this as an exploratory phase where we're looking for cleanest starts of a minimum viable product.
> 
> Ideally, we reproduce this in "pure" python: https://github.com/sotanakamura/winui3-without-xaml
> 
> I'll be fiddling presently, so feel free to send examples of what you have so far.

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\johnd\AppData\Roaming\Python\Python312\site-packages\win32com\client\__init__.py", line 118, in Dispatch
    dispatch, userName = dynamic._GetGoodDispatchAndUserName(dispatch, userName, clsctx)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\johnd\AppData\Roaming\Python\Python312\site-packages\win32com\client\dynamic.py", line 104, in _GetGoodDispatchAndUserName
    return (_GetGoodDispatch(IDispatch, clsctx), userName)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\johnd\AppData\Roaming\Python\Python312\site-packages\win32com\client\dynamic.py", line 86, in _GetGoodDispatch
    IDispatch = pythoncom.CoCreateInstance(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
pywintypes.com_error: (-2147221005, 'Invalid class string', None, None)

**Comment by jhi2** [comment]
Is there any way to fix this

**Comment by JPHutchins** [comment]
> Is there any way to fix this

I was hoping that you already had 🤣!  I am seeing the same thing, and I suspect that it's (one of) the reason that we're being steered towards alternatives by zooba.

**Comment by JPHutchins** [comment]
@jhi2 Found this little tool: `python -m win32com.client.combrowse`

**Comment by JPHutchins** [comment]
I have built the example applications from https://github.com/zooba/pymsbuild-winui

Unfortunately I didn't take notes, but after some fiddling, I'd suggest the following workflow:
* add `.venv` to the .gitignore
* `python -m venv .venv`
* `. .venv/scripts/activate`
* pip install .
* pip install -r tests/requirements.txt
* `cd tests`
* `pytest`
* I did get a working executable at `pymsbuild-winui\tests\testdata\app1\build\bin\TestApp\app.exe`

Other thoughts
* (PhotoViewer exe didn't work - it didn't move the python dlls and stuff into the app folder for some reason)
* build is putting a lot of files in tree instead of inside the build or temp folder which is annoying
* the TestApp folder is 228 MB - might be good though - the microsoft DLLs are getting copied over

Next steps
* Create a package using pymsbuild that provides a hello world to some single WinUI3 API call

**Comment by jhi2** [comment]
> @jhi2 Found this little tool: `python -m win32com.client.combrowse`

Nice!

**Comment by jhi2** [comment]
have you seen https://pypi.org/project/winrt/

**Comment by jhi2** [comment]
look at https://pypi.org/project/winrt/

**Comment by freakboy3742** [comment]
@jhi2 The Python WinRT projection is the subject of [literally the first comment on the WinUI3 thread](https://github.com/microsoft/microsoft-ui-xaml/issues/3364#issuecomment-703001044). AIUI, it can't be used for exposing WinUI3 APIs because [you can't build an automated binding to C++](https://github.com/microsoft/microsoft-ui-xaml/issues/3364#issuecomment-1826343042).

**Comment by jhi2** [comment]
have you seen:

1. https://pywinrt.readthedocs.io/en/latest/api/microsoft.html
2. https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/

**Comment by freakboy3742** [comment]
Can't say I've seen those. Last time I looked, winrt didn't have any GUI bindings at all; and from my discussions with people at MS, it wasn't plausible to add them, either. If that isn't the case any more, then it might be possible to build a WinUI3 backend using winrt as the binding.

**Comment by JPHutchins** [comment]
> have you seen:
> 
> 
> 
> 1. https://pywinrt.readthedocs.io/en/latest/api/microsoft.html
> 
> 2. https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/

This was my original starting point and was unable to replicate the win32more example.

**Comment by jhi2** [comment]
Hey, for my app, can you make a win32more-like binding for my app for pymsbuild-winui. submit as pr to jhi2/PyUI3

**Comment by freakboy3742** [comment]
@jhi2 I'm not looking for more projects to contribute to. I'd certainly *like* a WinUI3 binding to exist, but it's not a priority for me at present, partially because Toga *has* a binding that works on Windows, and partially because I don't use Windows as a platform myself. If you want a WinUI3 binding, you'll need to either develop it yourself, pay for someone else to develop it for you, or wait for someone else to develop it for you.

**Comment by JPHutchins** [comment]
> Hey, for my app, can you make a win32more-like binding for my app for pymsbuild-winui. submit as pr to jhi2/PyUI3

Go ahead and tag me on a discussion or issue on that repo to continue this discussion.  I should not have criticized win32more so much... it does seem to be providing a programmatic interface to WinUI3 which is more than other attempts have yielded.  Perhaps it's complexity is necessary and something that can be improved over time.  Anyway, keep going with that and see what you can do!  But let's work outside of Toga until there's something to offer Toga.

**Comment by jhi2** [comment]
Yeah, an error. I'll tag u.

**Comment by KubaO** [comment]
> > Can you look it over?
> 
> I can look it over. Think of this as an exploratory phase where we're looking for cleanest starts of a minimum viable product.
> 
> Ideally, we reproduce this in "pure" python: https://github.com/sotanakamura/winui3-without-xaml
> 
> I'll be fiddling presently, so feel free to send examples of what you have so far.

pywin32more makes it a solved problem. It's an impressive Python binding I must say.

```
from win32more.xaml import XamlApplication
from win32more.Windows.Foundation import Uri
from win32more.Microsoft.UI.Xaml import (
    HorizontalAlignment,
    ThicknessHelper,
    VerticalAlignment,
    Window,
    )
from win32more.Microsoft.UI.Xaml.Controls import (
    Button,
    HyperlinkButton,
    StackPanel,
    TextBlock,
    )
from win32more.Microsoft.UI.Xaml.Media import (
    DesktopAcrylicBackdrop,
    )


class App(XamlApplication):
    def OnLaunched(self, args):
        window = Window()
        self.window = window
        window.SystemBackdrop = DesktopAcrylicBackdrop()
        
        stackPanel = StackPanel()
        stackPanel.HorizontalAlignment = HorizontalAlignment.Center
        stackPanel.VerticalAlignment = VerticalAlignment.Center

        title = TextBlock()
        title.Style = App.Current.Resources.Lookup("TitleTextBlockStyle")
        title.Text = "WinUI 3 in C++ Without XAML!"
        title.HorizontalAlignment = HorizontalAlignment.Center
        
        project = HyperlinkButton()
        project.Content = "Github Project Repository"
        project.NavigateUri = Uri("https://github.com/sotanakamura/winui3-without-xaml")
        project.HorizontalAlignment = HorizontalAlignment.Center

        button = Button()
        button.Content = "Click Me"
        button.Click += lambda sender, args: button.put_Content("Thank You!")
        button.HorizontalAlignment = HorizontalAlignment.Center
        button.Margin = ThicknessHelper.FromUniformLength(20);
        
        window.Content = stackPanel
        stackPanel.Children.Append(title)
        stackPanel.Children.Append(project)
        stackPanel.Children.Append(button)
        
        window.Activate()
        
    def ButtonClick(self, sender, args):
        sender.as_(Button).Content = "Thank You!"


if __name__=="__main__":
    XamlApplication.Start(App)
```

**Comment by KubaO** [comment]
> > I would just want to call C++ code from ctypes. I found a way to call Python from C/C#/C++(Python.h)
> 
> This might be nice, but I'll recommend pymsbuild-winui or win32more for generating the bindings. I think it might be possible with ctypes but feels hacky. I mean, it's all hacky. That RoGetActivationFactory will provide a function pointer, but I'm not sure how to bind it to a python function. Maybe with this: https://docs.python.org/3/library/ctypes.html#callback-functions

win32more uses ctypes only. It's pure Python, machine-generated using a generator written in Python from JSON API description available elsewhere. Pretty much all of Windows API that is available from C is covered, and WinUI is definitely available via C, since COM works in C just fine, albeit it's a pain to use without generated code.

**Comment by KubaO** [comment]
> But having now implemented 4 different approaches of accessing it from Python, the one I think is best is to just generate C++ code to do all the GUI side and have it call back into Python.

WinUI and WinRT have C++ projections, they are not C++-native. The C++ code drops down to good old COM, so there's little benefit to the C++ shim if we want to use it from Python. C# does the same. We might as well use the same COM interfaces that the C++ projection uses, but expose them Pythonically as Python classes etc.

And that's what pywin32more does - it uses low-level COM directly, via ctypes. And it exposes the non-COM Win32Api as well. Since it's machine-generated, it provides all the constants, structures and functions that the C Win32 headers provide. Using it feels fairly natural to me.

At the moment, there is no other Python projection of Windows APIs that is competitive IMHO.

**Comment by freakboy3742** [comment]
Tagging @th3w1zard1 on this; it looks like pywin32more might implement the sort of reflective binding that we've been discussing in the context of #2786. That *might* make it a better match than comtypes for that work, while also leaving the door open for a full WinUI3 backend.

**Comment by freakboy3742** [comment]
> From what I've researched, yes pywin32 is a solution and can directly replace comtypes.

To clarify - [`pywin32`](https://pypi.org/project/pywin32/) and [`py-win32more`/`win32more`](https://pypi.org/project/win32more/) are two different projects. The discussions on this ticket have been about the latter.

**Comment by jhi2** [comment]
> > From what I've researched, yes pywin32 is a solution and can directly replace comtypes.
> 
> To clarify - [`pywin32`](https://pypi.org/project/pywin32/) and [`py-win32more`/`win32more`](https://pypi.org/project/win32more/) are two different projects. The discussions on this ticket have been about the latter.

My COM post was about the former.

**Comment by th3w1zard1** [comment]
> > > From what I've researched, yes pywin32 is a solution and can directly replace comtypes.
> > 
> > 
> > To clarify - [`pywin32`](https://pypi.org/project/pywin32/) and [`py-win32more`/`win32more`](https://pypi.org/project/win32more/) are two different projects. The discussions on this ticket have been about the latter.
> 
> My COM post was about the former.

Yeah exactly... so what am I missing here?

Could we talk about what is happening to this and #2786 ...? I'd like to know if it'd be worthwhile to continue holding off with `comtypes` if we believe something better could be implemented here?

**Comment by KubaO** [comment]
py-win32more is a projection of the full Windows API. It includes pretty much everything documented by MS. It’s not just a way to consume winrt api. It is a Python way of using Windows. I am using it professionally. It covers everything that comes in Windows SDK headers. It also has runtime type checking that helps avoid crashes that the dynamicism of Python would otherwise make easy to run into.pywin32 is a small projection of WinAPI - it covers a small part of it, maybe 20%? comtypes is just a framework one could use to consume winrt (with some pain). You’d still be typing in the signatures of stuff.As far as I am concerned, win32more is THE way to use Windows APIs in Python. There’s another professional projection of WinAPI that includes winrt stuff - Microsoft’s own winrt package on pip. Caveat emptor as MS had a tendency of dropping OSS projects randomly.If win32more’s strength is the simplicity of implementation, complete coverage of the APIs, and ease of vendoring if ever needed, winrt’s strength is the example code - there is a lot of it, and it’s pretty good. In most cases, „translating” it to win32more is trivial and takes changing the imports. Other minor adjustments may be needed too.I have more trust in win32more’s longevity than winrt’s. Should the worst happen, I am familiar enough with it to continue maintaining it. I have a fairly large internal project that uses it.I’ve looked at how much work it would be to get a proof of concept WinUI 3 backend for toga that can get at least a few of the more complex examples working. I’ll now shut up and work on that instead. Expect a PR for the PoC so that we can at least have something concrete to talk about :)2. nov. 2024 kl. 5:22 am skrev Benjamin Auquite ***@***.***>:﻿



From what I've researched, yes pywin32 is a solution and can directly replace comtypes.

To clarify - pywin32 and py-win32more/win32more are two different projects. The discussions on this ticket have been about the latter.

My COM post was about the former.

Yeah exactly... so what am I missing here?
Could we talk about what is happening to this and #2786 ...? I'd like to know if it'd be worthwhile to continue holding off with comtypes if we believe something better could be implemented here?

**Comment by freakboy3742** [comment]
> > > > From what I've researched, yes pywin32 is a solution and can directly replace comtypes.
> > > 
> > > 
> > > To clarify - [`pywin32`](https://pypi.org/project/pywin32/) and [`py-win32more`/`win32more`](https://pypi.org/project/win32more/) are two different projects. The discussions on this ticket have been about the latter.
> > 
> > 
> > My COM post was about the former.
> 
> Yeah exactly... so what am I missing here?
 
I'm not sure who this question is directed at - but personally, I have no idea why pywin32 is being mentioned at all. 

> Could we talk about what is happening to this and #2786 ...? I'd like to know if it'd be worthwhile to continue holding off with `comtypes` if we believe something better could be implemented here?

From my personal perspective - *nothing* is happening here. This ticket exists to flag the fact that yes, we'd like a WinUI3 backend. It's not on my own roadmap, nor is it likely to be in the near future. However, if someone were to contribute a WinUI3 backend, I'd look at it. 

To that end, the fundamental issue that needs to be resolved at this point is the interface that will be used to implement the Python-WinUI3 bridge. The Winforms backend uses python.net. That provides a way to invoke Winforms APIs from Python in a format where it's obvious how to read official Winforms documentation and turn that into Python code. I don't use Windows on a daily basis, and I'm not deeply familiar with Windows GUI programming, but I can use Python.net to read Winforms documentation and implement Toga's APIs. What is the equivalent for WinUI3? Based on what has been said in this thread, win32more sounds (to me) like the most viable option - but I'm saying that on the basis of no practical experience, and a single example app presented above. I haven't explored what (if any) limits the win32more approach has when it comes to more complicated widgets, the stability/maintainablity of the win32more project, or whether it addresses any of the issues that currently face the winforms backend (e.g., support for ARM architectures, support for widgets like DetailedList or Tree, or modern dialog support).

As for #2786 specifically - nothing has changed since my last comment. The most pressing issue is that the testbed tests are currently failing. That needs to be resolved. I've already provided some pointers; if you need something more specific to get unstuck then you'll need to indicate where you need help.

**Comment by johnzhou721** [comment]
> I haven't explored what (if any) limits the win32more approach has when it comes to more complicated widgets, the stability/maintainablity of the win32more project, or whether it addresses any of the issues that currently face the winforms backend (e.g., support for ARM architectures, support for widgets like DetailedList or Tree, or modern dialog support).

ARM64 support -- we can see some code handling ARM64 like https://github.com/ynkdir/py-win32more/blob/81ed9c0819e0b72602a576b5e4e5881e1cfe98d6/src/win32more/appsdk/mddbootstrap.py#L44-L46 but not sure if it'd work properly.  Running ``python script.py`` on ARM64 does work (after modifying 1 file on installed win32more which I'll mention later) but I'm not sure if it's getting emulated somewhere.

There is https://github.com/ynkdir/py-win32more/issues/111 though which I patched the _IsWin11 function to uncomment the commented out line and not return False all the time on the relevant file linked in that thread.  May be a good test for stability/maintainability to see if the author responds really fast.  Maintainer's very responsive though, last commit is 5 days ago as of writing now even though it's been like a few months since any discussion here.

Tree widget -- seems like the API is accessible, don't know about any quirks, tested with code generated by ChatGPT (that mostly makes sense to me with no attempt to construct its own Tree widget):

<details>
  <summary>Code</summary>

```
from win32more.xaml import XamlApplication
from win32more.Microsoft.UI.Xaml import (
    HorizontalAlignment,
    VerticalAlignment,
    Window,
    ThicknessHelper,
)
from win32more.Microsoft.UI.Xaml.Controls import (
    StackPanel,
    TextBlock,
    Button,
    TreeView,
    TreeViewNode,
)
from win32more.Microsoft.UI.Xaml.Media import DesktopAcrylicBackdrop


class App(XamlApplication):
    def OnLaunched(self, args):
        window = Window()
        self.window = window
        window.SystemBackdrop = DesktopAcrylicBackdrop()

        stack = StackPanel()
        stack.HorizontalAlignment = HorizontalAlignment.Center
        stack.VerticalAlignment = VerticalAlignment.Center

        title = TextBlock()
        title.Text = "WinUI3 TreeView (TreeViewNode) in Python"
        title.HorizontalAlignment = HorizontalAlignment.Center
        stack.Children.Append(title)

        # Create TreeView (node-based)
        tree = TreeView()
        tree.HorizontalAlignment = HorizontalAlignment.Center
        tree.VerticalAlignment = VerticalAlignment.Center

        # Create root node 1
        root1 = TreeViewNode()
        root1.Content = "Root 1"

        # Children for root1
        c11 = TreeViewNode()
        c11.Content = "Child 1.1"
        c12 = TreeViewNode()
        c12.Content = "Child 1.2"
        root1.Children.Append(c11)
        root1.Children.Append(c12)

        # Create root node 2
        root2 = TreeViewNode()
        root2.Content = "Root 2"
        c21 = TreeViewNode()
        c21.Content = "Child 2.1"
        root2.Children.Append(c21)

        # Add roots to the TreeView
        # WinUI exposes RootNodes (not Items/Items.Add for nodes)
        tree.RootNodes.Append(root1)
        tree.RootNodes.Append(root2)

        # Handle when a node is invoked (clicked/activated)
        def on_item_invoked(sender, args):
            # args.InvokedItem is the underlying invoked item (the node/content)
            invoked = getattr(args, "InvokedItem", None)
            # In many cases InvokedItem is the node's Content - print it
            print("Invoked item:", invoked)

        tree.ItemInvoked += on_item_invoked

        stack.Children.Append(tree)

        btn = Button()
        btn.Content = "Click Me"
        btn.Margin = ThicknessHelper.FromUniformLength(12)
        btn.Click += lambda s, a: btn.put_Content("Thanks!")
        stack.Children.Append(btn)

        window.Content = stack
        window.Activate()


if __name__ == "__main__":
    XamlApplication.Start(App)
```

</details>

<img width="164" height="165" alt="Image" src="https://github.com/user-attachments/assets/d15f82de-a021-452d-867f-85c804bd4d0f" />

Busy with a lot of other things I have no intents to work on this but it seems that it's possible to access some more complicated things with win32more.  Just trying to knock out some preliminary research for y'all showing more viability.

**Comment by johnzhou721** [comment]
Heads up:  py-win32more just did a restructuring 1 week after posting of the previous comment that will hopefully resolve some of y'alls code organization concerns.

Right now at the root of https://github.com/ynkdir/py-win32more/tree/main there is the binding generator tool at win32generator; https://github.com/ynkdir/py-win32more/blob/main/scripts/build_subpackage.py and https://github.com/ynkdir/py-win32more/blob/main/.github/workflows/update_sdk.yml are used to build the actual packages; built bindings are stored inside https://github.com/ynkdir/py-win32more/tree/main/packages; the root win32more sources are just simply like glue code.

Finally the dependencies at https://github.com/ynkdir/py-win32more/blob/main/pyproject.toml specifies the bindings as the specific dependencies in a way similar to pyobjc

All of this should resolve @JPHutchins ' concerns here
> Generally, I am having a hard time understanding that repo, it is very complex. It sorta looks like a code generation tool PLUS some generated code, which makes usage confusing.

The code generator is in the locations I talked about, and since there's a github actions now for building bindings how the bindings gets generated should now be completely transparent.

**Comment by johnzhou721** [comment]
OK... seems that win32more also provides asyncio integration.  Nice!  There's a few issues with it (like pumping only once per 100ms) that I'd probably want to file upstream when I get around to working on this.

I plan to start a WinUI3 backend after I port more widgets to GTK4 and sort out some Qt backend things.  But if anyone else wants to get started, go ahead and open a PR here (this seems pretty viable to me).

**Comment by johnzhou721** [comment]
I'm dropping any thing I offer to make a WinUI3 backend, due to personal things.

Anyways:  It seems that pywinrt, which is a fork of the official WinRT xlang project by Microsoft for Python, is also exposing the Xaml component.  It's an alternative to pywin32more, it was forked off of an official project by Microsoft.  I haven't ran it yet, but it seems that it also exposes the Xaml Application facilities:  https://github.com/pywinrt/pywinrt/blob/dcb06e3871dd26ffddbdf7d06a07a7d5b73610af/projection/winrt/winrt-Windows.UI.Xaml/winrt/_winrt_windows_ui_xaml.pyi#L128-L221

**Comment by Oliver-Leigh** [comment]
I've decided to get the ball rolling on this feature. If someone else is working on this let me know!

As far as I can see both options discussed (win32more and pywinrt) can be used to build this feature. For my own reference, here's a brief overview of each:

- pywinrt: Built in C++ from an official Microsoft repository. It's an impressive project and feels very clean/professional and has nice documentation. It's also had some pretty senior Microsoft staff as contributors. The downside here is that I'm not sure how actively it's being developed. It seems stuck moving from App SDK 1.7 to 1.8 and the last 3 dates for commits are 11-Dec-2025, 04-Sep-2025 and 06-Jun-2025. Also, as it's written in C++ it's a lot less transparent for Toga developers to use. 

- win32more: Built in python (although the metadata extraction tool winmd-printer is built in C#) using ctypes to access the dll files. It's essentially built by one contributor and is actively being developed. I think win32 is awesome, but it has some downsides: 
  
    - There is essentially no documentation.
    - The code base is very hard to audit (for me). There is minimal commenting/documentation and some near circular references. 
    - External dlls are hardcoded into the repository

@freakboy3742 Before I invest too much time in this, I'd like to know your thoughts here based on what I written above. I see three options for moving forward: Using one of the two above repositories as is, or using a hybrid approach where some of the code from win32more is rewritten and incorporated directly into Toga. For example, the core interface of win32more has a pretty small codebase and so could be completely incorporated. My thought here is that you'd like to avoid a python.net situation, but this will also create more upkeep.

**Comment by mhsmith** [comment]
When you refer to pywin32, do you mean pywinrt?

**Comment by Oliver-Leigh** [comment]
Oops. Yes, that's what I meant. I updated the comment.

**Comment by johnzhou721** [comment]
**Below, winrt refers to the PyWinRT project maintained and further developed by the community**.

Well... pywinrt isn't actively developed, that's a point.  But also:
1. Large sections of the docs is essentially missing; the rest is mostly documenting how the conventions of how the API is wrapped

<img width="384" height="861" alt="Image" src="https://github.com/user-attachments/assets/9747d887-dbde-4bb9-a6d4-c44df7851102" />

2. WinRT's interaction with packaging has not been fully explored (see the API Reference section).  With win32more even though we're still in the "unofficial workarounds" territory, these types of issues when filed have been resolved in 48 hours or less: https://github.com/ynkdir/py-win32more/issues/111
3. So with win32more it's possible to get more documentation and explanations as issues come up.  With WinRT it seems that it's a small set of docs covering the naming conventions etc. of the API, some well-documented code, but if things come up we may lose the abiilty to get ongoing support.
4. WinRT has *way* more dependents than win32more: https://libraries.io/pypi/winrt-runtime, https://libraries.io/pypi/win32more, so WinRT seems mroe reliable in this regard

Docs issue filed as https://github.com/ynkdir/py-win32more/issues/176

**Comment by Oliver-Leigh** [comment]
@johnzhou721 Good points, but can you clarify where you mean PyWinRT and where you mean WinRT?

**Comment by johnzhou721** [comment]
> [@johnzhou721](https://github.com/johnzhou721) Good points, but can you clarify where you mean PyWinRT and where you mean WinRT?

I mean pywinrt all the way through.  I've edited my comment to reflect this.

When you're here, mark your comment as resolved to follow precedent :)

**Comment by freakboy3742** [comment]
> I've decided to get the ball rolling on this feature. If someone else is working on this let me know!

🎉 

I'm not aware of anyone working on this - so go right ahead.

> * pywinrt: Built in C++ from an official Microsoft repository. It's an impressive project and feels very clean/professional and has nice documentation. It's also had some pretty senior Microsoft staff as contributors. The downside here is that I'm not sure how actively it's being developed. 

Everything I've heard from people inside MS is that, for all intents and purposes, it *isn't* being maintained. There hasn't been a release since 2021, and there's barely been commit activity in a year. Even when it was being actively worked on, the focus was wrapping ML classes; AIUI, there's some fundamental limitation to wrapping GUI classes that they weren't especially motivated to resolve.

> * win32more: Built in python (although the metadata extraction tool winmd-printer is built in C#) using ctypes to access the dll files. It's essentially built by one contributor and is actively being developed. I think win32 is awesome, but it has some downsides: 
>   * There is essentially no documentation.
>   * The code base is very hard to audit (for me). There is minimal commenting/documentation and some near circular references.
>   * External dlls are hardcoded into the repository

I guess the question for me is whether the one contributor is looking for assistance. It's possible the lack of documentation and code quirks like hardcoded DLLs and circular references are a matter of expediency/practicality; and they might be receptive to refactors and other improvements. The fact that there's no open PRs, a lot of the open tickets are questions rather than "doesn't work" style bugs, and the speed of response to some of the questions asked suggests they might be open to contributions.
 
> [@freakboy3742](https://github.com/freakboy3742) Before I invest too much time in this, I'd like to know your thoughts here based on what I written above. 
>
> I see three options for moving forward: Using one of the two above repositories as is, or using a hybrid approach where some of the code from win32more is rewritten and incorporated directly into Toga. For example, the core interface of win32more has a pretty small codebase and so could be completely incorporated. My thought here is that you'd like to avoid a python.net situation, but this will also create more upkeep.

If pywinrt were being actively maintained by Microsoft, it would be an easy choice - but given its current state, I think it's a non-starter.

That leaves win32more. I'd definitely like to avoid a repeat of the python.net situation as well - but the alternative of taking on ownership of language binding isn't *that* much more appealing. But if that is the best option, I wouldn't rule it out. 

Incorporating the core of win32more into a Toga backend *might* be workable, depending on how much code is involved. My main hesitation on that front is maintainability. If the code is complex enough that it's difficult for *us* to maintain it; or we're likely to end up in a situation where there's enough useful and potentially reusable code that a third party wants to re-use it, then I'd feel awkward including it in Toga's codebase. 

I guess my immediate preference would be to see whether the win32more maintainer is easy to collaborate with. If they're responsive to feedback and fixes, then using it as a third party doesn't sound too unworkable.

**Comment by JPHutchins** [comment]
> it isn't being maintained. There hasn't been a release since 2021, and there's barely been commit activity in a year.

Last release was less than a year ago: https://github.com/pywinrt/pywinrt/releases

I am grateful for @dlech's maintenance of that lib. It is a dependency of Bleak's windows backend.

**Comment by freakboy3742** [comment]
My apologies - I've clearly got myself tied in knots over `winrt` vs `pywinrt` (and the naming of the package on PyPI confuses matters even further).

Now that I'm looking at the right package, `pywinrt` definitely seems a lot more viable. There hasn't been a lot of activity in the last year... but if the bulk of the code is automatically generated language projections, maybe there isn't a lot of need for activity? It's definitely a good sign that after a July release, they pushed additional packages in September for Py3.14 support.

**Comment by Oliver-Leigh** [comment]
Thanks for the thoughts! 

Since they're both projections of the same api, maybe it doesn't matter so much at this stage which library is used? It might be possible to change between the two before the codebase becomes too big.

I think I'll make a choice and get to work on making a minimal viable implementation.

**Comment by johnzhou721** [comment]
> My apologies - I've clearly got myself tied in knots over `winrt` vs `pywinrt` (and the naming of the package on PyPI confuses matters even further).
> 
> Now that I'm looking at the right package, `pywinrt` definitely seems a lot more viable. There hasn't been a lot of activity in the last year... but if the bulk of the code is automatically generated language projections, maybe there isn't a lot of need for activity? It's definitely a good sign that after a July release, they pushed additional packages in September for Py3.14 support.

There *is* a need for activity—it's been 5 months and there is a need to update fore the latest windows SDK: https://github.com/pywinrt/pywinrt/issues/145 — @JPHutchins will you perhaps be able to work on this update so it can stay up to date?  PyWinRT is already 2 versions behind after 2.0.1 of Windows SDK released in 2026 January; win32more uses GitHub actions to automatically update SDKs as soon as they come out (2.0 is released just yesterday—

Another issue is size.  I've asked ChatGPT to count the sizes for win32more, and I provided it a list of win32more subpackages, and then tweaked the scirpt myself bit to count the pure python wheel and the most recent version only, and `win32more` totals to 9.49MB; but if I run the same script on only *a few* out of the many pywinrt components — (Microsoft.UI.Xaml, WindowsAppSDK, runtime, sdk)— (since it's more modular), we're already at 7.74MB (see the dtails section for the scripts).

So win32more seems to be less modular, but it overall saves a lot of size.  Excluding the 3.85MB Win32Metadata module, where we can probably use ctypes manually, *all* of the bindings of win32more are going to be smaller than even just a few core components of pywinrt.

<details>

```
import requests

packages = [
    "win32more",
    "win32more-appsdk",
    "win32more-core",
    "win32more-Microsoft.WindowsAppSDK",
    "win32more-Microsoft.Graphics.Win2D",
    "win32more-Microsoft.Windows.SDK.Win32Metadata",
    "win32more-Microsoft.Web.WebView2",
    "win32more-Microsoft.Windows.SDK.Contracts",
]

def human(n):
    for unit in ["B", "KB", "MB", "GB"]:
        if n < 1024:
            return f"{n:.2f} {unit}"
        n /= 1024

grand_total = 0

for pkg in packages:
    url = f"https://pypi.org/pypi/{pkg}/json"
    r = requests.get(url)
    
    if r.status_code != 200:
        print(f"{pkg}: FAILED")
        continue

    data = r.json()
    latest = data["info"]["version"]
    files = data["releases"].get(latest, [])

    wheels = [f for f in files if f["filename"].endswith(".whl")]

    chosen = None

    for w in wheels:
        if "py3-none-any" in w["filename"]:
            chosen = w
            break

    size = chosen["size"] if chosen else 0
    grand_total += size

    print(f"{pkg:55} {human(size)}")

print("\nTOTAL:", human(grand_total))
```

```
johnzhou@Johns-MacBook-Pro ~ % python -m win321.py
win32more                                               3.58 KB
win32more-appsdk                                        11.86 KB
win32more-core                                          25.24 KB
win32more-Microsoft.WindowsAppSDK                       1.17 MB
win32more-Microsoft.Graphics.Win2D                      1.88 MB
win32more-Microsoft.Windows.SDK.Win32Metadata           3.85 MB
win32more-Microsoft.Web.WebView2                        842.86 KB
win32more-Microsoft.Windows.SDK.Contracts               1.73 MB

TOTAL: 9.49 MB
```

```
johnzhou@Johns-MacBook-Pro ~ % python -m win322.py
winrt-Microsoft.UI.Xaml                                           221.61 KB
winrt-WindowsAppSDK                                               1.60 MB
winrt-runtime                                                     406.47 KB
winrt-sdk                                                         5.53 MB

TOTAL: 7.74 MB
```
</details>


@Oliver-Leigh I've made the initial implmeentaion for the Qt backend of Toga; let me know if you need some tips to get up and running.  I'm also happy to collaborate on code in any way we deem appropriate, if I get some time.

**Comment by freakboy3742** [comment]
> ... will you perhaps be able to work on this update so it can stay up to date? 

@johnzhou721 Please be *very* careful about requesting that other people do work. Asking someone to work on something when they haven't offered to do so, and you're not offering to pay them, is quite rude.

**Comment by johnzhou721** [comment]
> > ... will you perhaps be able to work on this update so it can stay up to date?
> 
> [@johnzhou721](https://github.com/johnzhou721) Please be _very_ careful about requesting that other people do work. Asking someone to work on something when they haven't offered to do so, and you're not offering to pay them, is quite rude.

Thanks for letting me know of this consideration.

@JPHutchins I apologize for the wording of this statement, and the fact that it might've came off as rude.  Don't feel pressured to update anythignn unless you're motivated to do so.

**Comment by zooba** [comment]
> AIUI, there's some fundamental limitation to wrapping GUI classes that they weren't especially motivated to resolve.

It's inheritance. The GUI model fundamentally relies on inheritance hierarchies, and is intended for the developer to create their own subclasses overriding protected methods. The mechanics for this are pretty unique to WinRT, and Python would be the first non-compiled language to support them (if you do it properly - I think you can manage basic GUIs or incredibly hacky ones without it, or possibly more since I last looked). Most of the rest of WinRT doesn't rely on it at all.

I'll admit, I had a bit of fun implementing them last time I did it. It was just never ready to release, unfortunately.

**Comment by freakboy3742** [comment]
> > AIUI, there's some fundamental limitation to wrapping GUI classes that they weren't especially motivated to resolve.
> 
> It's inheritance. The GUI model fundamentally relies on inheritance hierarchies, and is intended for the developer to create their own subclasses overriding protected methods. The mechanics for this are pretty unique to WinRT, and Python would be the first non-compiled language to support them (if you do it properly - I think you can manage basic GUIs or incredibly hacky ones without it, or possibly more since I last looked). Most of the rest of WinRT doesn't rely on it at all.
> 
> I'll admit, I had a bit of fun implementing them last time I did it. It was just never ready to release, unfortunately.

Thanks for that detail @zooba. 

Should I take it from this that you've got as far as a proof-of-concept for inheritance support, but nothing complete/stable enough that it's releasable? If so - any chance you could share your incomplete implementation (or provide hints in the direction that someone else should take if they wanted to replicate your work)?

**Comment by zooba** [comment]
> Should I take it from this that you've got as far as a proof-of-concept for inheritance support, but nothing complete/stable enough that it's releasable?

I've attempted this project four times since 2011, but they've all been on work time so nothing I can share. The pywinrt/winrt package is the best so far (they actually spoke to me early and avoided a big range of problems), though I wasn't involved in it.

My current best idea is [pymsbuild-winui](https://github.com/zooba/pymsbuild-winui), which basically lets you define the UI using XAML and compile it to a native executable host that then calls back into your Python code. That goes against "everyone's" preference for dynamically building up a GUI, but that idea also failed the first three times I tried it so you'll forgive me for not trying that approach again :)

**Comment by zooba** [comment]
As for hints about implementing inheritance, just make sure you understand COM's interface model _really_ well (i.e. you can implement a multi-interface COM class in C without scaffolding or reference material). Once that's clear in your head and [examples like this](https://www.interact-sw.co.uk/iangblog/2011/09/25/native-winrt-inheritance) stop being scary, you've got a chance 😉

**Comment by johnzhou721** [comment]
> pywinrt/winrt

Did you mean the GitHub repo pywinrt/pywinrt (winrt-runtime on PyPI), or do you mean either PyWinRT or Microsoft's PyWinRT from xlang?  Do you think PyWinRT provides any benefits over the older Microsoft xlang regarding inheritance issues?  Are there any other inheritance issues in PyWinRT you are aware of?  @zooba — Thanks.

**Comment by zooba** [comment]
> Did you mean the GitHub repo pywinrt/pywinrt (winrt-runtime on PyPI), or do you mean either PyWinRT or Microsoft's PyWinRT from xlang?

Pretty sure it's all the same project, just one is forked and maintained from the other. I've stopped tracking them closely, so feel free to dig into whichever one is being updated and use that.

> Do you think PyWinRT provides any benefits over the older Microsoft xlang regarding inheritance issues?

Last I looked nobody even attempt to implement it. The critical piece being to implement functions declared virtual in a WinRT base class with an overriding implementation in Python. It's essentially just a callback, but it's structured more like dependency injection than an event handler. Very clever design, possibly a bit too clever, but it was never meant to be implemented over and over again - the intention at the time was for .NET and a C++ template library to support it, and that was it (JavaScript wasn't going to get support, IIRC, though I'm sure the team looked at it and possibly influenced it in a way that made it more feasible for a dynamic language to play a part).

**Comment by dlech** [comment]
> Do you think PyWinRT provides any benefits over the older Microsoft xlang regarding inheritance issues?

Yes. The old xlang Python projection didn't implement it. It is implemented now in the fork I maintain. I don't know if anyone has actually tried using it much though. It works in some basic examples. win32more is probably more battle-tested in that area.

**Comment by johnzhou721** [comment]
Another datapoint about win32more:  It seems that upgrading may be unreliable, refs https://github.com/ynkdir/py-win32more/issues/180 — not sure if this is a general quirk or something with my environment.

Previously I said win32more was very fast to update to new SDKs, however there's some bugs in the process: see e.g. https://github.com/ynkdir/py-win32more/issues/179 — so my previous statement about a fast update wasn't particularly accurate.

But the maintainers do seem very responsive... so new bugs can probably get fixed quickly.

**Comment by Oliver-Leigh** [comment]
I’ve implemented a proof of concept [here](https://github.com/Oliver-Leigh/toga/tree/winui-3/winui3/src/toga_winui3). There is enough written so that the `Button` example can be run. 

To run this you’ll need the Windows App SDK components installed. I haven’t looked into the simplest way to do this yet, but I installed Visual Studio with the  C++ WinUI app development tools (as described [here](https://learn.microsoft.com/en-us/windows/apps/get-started/start-here?tabs=manual)), and also the runtime from [here](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/downloads).

**Known issues:**
If something's not working, it's probably because I haven't implemented it yet. However, there are two things I haven't worked out yet. 
 - How do external resources work?  I haven’t been able to implement icons fully because of this. I can do a work around through the use of a Win32 stream, but I’d rather not. Any help here would be great! 
 - There is an issue shutting down the main loop. This is most easily seen when exiting the app through the main window. I haven't spent time debugging this yet because I'll be rewriting the proactor loop soon to get rid of the imposed performance delay. 

**Moving forward:**
I don’t see major hurdles for the remaining widgets, except Table and Tree. For some reason there is no built-in WinUI 3 widget for displaying multicolumn data. I’m not sure what the best option will be, but we can always fall back on the Win32 ListView through the use of an interop. 

The need to have the Windows App SDK installed is pretty annoying, but (if it is desirable) I think that it will be possible to configure Briefcase to package the necessary dlls. Something similar is already available on win32more. 

@freakboy3742 What level of completeness would you like to see for a PR base on this?

<img width="395" height="247" alt="Image" src="https://github.com/user-attachments/assets/a315c9ee-0836-4bd1-848e-6427c3b7ee1c" />

<img width="395" height="247" alt="Image" src="https://github.com/user-attachments/assets/064cf49f-af19-4eb5-9d8a-105a8ad9841c" />

**Comment by freakboy3742** [comment]
> I’ve implemented a proof of concept [here](https://github.com/Oliver-Leigh/toga/tree/winui-3/winui3/src/toga_winui3). There is enough written so that the `Button` example can be run.

🎉  Nice work!

> To run this you’ll need the Windows App SDK components installed. 

I'm a little hazy on what this would mean in practice. Is this a requirement for developers, or end users? Is it analogous to the "must install a .NET runtime" requirement for Winforms? Can we detect the case where the tools aren't available and provide a clear prompt for what needs to be installed? Is this something that an app installer can bundle? Would it require bundling a second MSI with the runtime with the main app?

> * How do external resources work?  I haven’t been able to implement icons fully because of this. I can do a work around through the use of a Win32 stream, but I’d rather not. Any help here would be great!

I'm not sure I'm going to be much use here... but what error/problem are you seeing?

> **Moving forward:** I don’t see major hurdles for the remaining widgets, except Table and Tree. For some reason there is no built-in WinUI 3 widget for displaying multicolumn data. I’m not sure what the best option will be, but we can always fall back on the Win32 ListView through the use of an interop.

... sigh. What does Microsoft have against multi-column lists?

> The need to have the Windows App SDK installed is pretty annoying, but (if it is desirable) I think that it will be possible to configure Briefcase to package the necessary dlls. Something similar is already available on win32more.

As noted above - that's my main question: what's the user experience for the person installing an app built with this backend?
 
> [@freakboy3742](https://github.com/freakboy3742) What level of completeness would you like to see for a PR base on this?

"Passes Testbed" (even if 95% of the test suite is "skip") is the minimal starting point for a PR. As long as *some* tests pass for at least *one* widget, that's all it needs to do initially.

**Comment by johnzhou721** [comment]
N.B.  The discussions at the end of https://github.com/ynkdir/py-win32more/issues/179; there’s clearly some coordination needed to ensure that the compatible version of windows app sdk is installed for the win32more component versions.  Explicit pins may be required

**Comment by Oliver-Leigh** [comment]
> I'm a little hazy on what this would mean in practice. Is this a requirement for developers, or end users?

The necessary dlls will need to be present on the hardware of both end users and developers. _However_, I envision them in different forms:

- **Developers** - Should install the whole Windows App SDK and the necessary dependencies. At the moment I've installed the Windows App SDK via Visual Studio (mainly because I wanted to explore the MS preferred solution), however this is not practical since the install size is ~20GB. I think the preferred solution for this will be to provide a script or instructions using the nuget packages from [here](https://www.nuget.org/packages/Microsoft.WindowsAppSDK), but I haven't worked out any details. 
- **End users** - Ideally, the required dlls will be supplied in the final package. We wouldn't need to include the whole Windows App SDK, just the necessary dlls. There is a [script](https://github.com/ynkdir/py-win32more/blob/main/scripts/build_selfcontained.py) on win32more which will be useful since it creates a stand alone deployment from the nuget packages.


> Is it analogous to the "must install a .NET runtime" requirement for Winforms?

Similar, but .NET provides a nice installer and there doesn't seem to be an equivalent for Windows App SDK. However, I _think_ it's going to be easier to package the required dlls directly with Windows App SDK. 

> Can we detect the case where the tools aren't available and provide a clear prompt for what needs to be installed?

Yes, this is possible. 
 
> Is this something that an app installer can bundle? Would it require bundling a second MSI with the runtime with the main app?

I think we can get away with one installer, but probably an MSIX package. Win32more provides an example of how to build such a package [here](https://github.com/ynkdir/py-win32more/blob/main/docs/howto-create-packaged-application.md).

**Comment by freakboy3742** [comment]
> > I'm a little hazy on what this would mean in practice. Is this a requirement for developers, or end users?
> 
> The necessary dlls will need to be present on the hardware of both end users and developers. _However_, I envision them in different forms:

If the issue is purely DLL distribution - as long as the DLLs are licensed in a way that permits distribution (and my read of the license says that it can be), is there any reason we can't treat those DLLs in a similar way to how we're handling the WebView DLLs? i.e., build a platform-specific wheel that includes the necessary DLLs, and load them from a known location at runtime?
 
> I think we can get away with one installer, but probably an MSIX package. Win32more provides an example of how to build such a package [here](https://github.com/ynkdir/py-win32more/blob/main/docs/howto-create-packaged-application.md).

FWIW - [MSIX installers are on my wish list for Briefcase](https://github.com/beeware/briefcase/issues/2097).

**Comment by Oliver-Leigh** [comment]
> I'm not sure I'm going to be much use here... but what error/problem are you seeing?

No problem! The issues is that I can't get the resources to load. I'll do a bit more of a deep dive here when the time is right. 

> If the issue is purely DLL distribution - as long as the DLLs are licensed in a way that permits distribution (and my read of the license says that it can be), is there any reason we can't treat those DLLs in a similar way to how we're handling the WebView DLLs? i.e., build a platform-specific wheel that includes the necessary DLLs, and load them from a known location at runtime?

Is this your preferred method for loading the DLLs? It would certainly help with the issue that @johnzhou721 raised. I think that it's possible. However there are two caveats: 

 1. There might be some issue with DLL lookup. We might have to coordinate with win32more to use explicit DLL locations. Otherwise a manifest might be required, and this would only make sense for Briefcase. 
 2. The Windows App SDK runtime may still need to be installed for developers. It may be avoidable with bootstrapping, but I haven't look into this yet. 

I’ll have to look more deeply into this and do some testing before I can give a definitive answer here.

**Comment by freakboy3742** [comment]
> > If the issue is purely DLL distribution - as long as the DLLs are licensed in a way that permits distribution (and my read of the license says that it can be), is there any reason we can't treat those DLLs in a similar way to how we're handling the WebView DLLs? i.e., build a platform-specific wheel that includes the necessary DLLs, and load them from a known location at runtime?
> 
> Is this your preferred method for loading the DLLs? 

"Preferred" implies I like it :-) ... but yes, based on my current understanding, I think this is the "least offensive" solution we have available to us. 

That said, this is definitely at the edge of my experience with Windows, so I'll absolutely defer to someone with a better idea. 

Coordination with win32more (or any similar project) should also be a broader goal - we should be aiming to make the whole ecosystem stronger, not just Toga.

**Comment by zooba** [comment]
Yes, that's the intended distribution method for the App SDK (and WebView). It's aiming to get people out of the "every app requires admin to install", which also means you need to vendor more of the app framework. (The idea is also that you're distributing an app that is significantly bigger than the framework, which isn't the case for a lot of Python stuff, and so it looks like a lot of files. It's intended for the Photoshops of the world, not the 100 line scripts of the world 😉 )

The ideal would be to include them with your app, rather than in a shared location, since a private copy means you don't have to coordinate shared versioning/updates (another major complaint from real developers). If you can scope down the group relying on a shared copy so they can all synchronise on updates, it's less of a problem. If you're going to offer a shared wheel then you'll want to be really clear about how to use it (e.g. don't pin the version, schedule CI tests against prerelease wheels, assume the world will change around your app, if you don't like it then vendor what you need).

**Comment by ynkdir** [comment]
Did you check https://github.com/microsoft/dynwinrt?
Maybe they are working on official projection.  It hasn't been released yet, though.

**Comment by JTSJohnny** [comment]
I appologize, I am the orginal author of this issue, Last year, i switched to Linux and have discontinued this implementation project.

**Comment by freakboy3742** [comment]
> Did you check https://github.com/microsoft/dynwinrt? Maybe they are working on official projection. It hasn't been released yet, though.

While this might eventually be an interesting library, it's currently 2 months and 160 commits old, with one committer, and no published PyPI package. It would be very premature to base anything on it.
